### PR TITLE
fix(clean flow cell api)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

Part of the project to implement new run devices. This PR addresses refactoring of the clean flow cells api.
Closes: https://github.com/Clinical-Genomics/add-new-tech/issues/29

### Changed
- Changed cli command `cg clean flow-cells` to `cg clean sequencing-runs`
- Renamed `CleanFlowCellApi` to `IlluminaCleanSequencingRunsService`
- Moved files related to the serves into `cg/services/illumina_services/illumina_cleaning_services`
- Moved testing suite into approriate places

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 1]
```diff
@@ -14,54 +14,63 @@
...
-        LOG.info(f"Trying to delete {flow_cell_path}")
+        LOG.info(f"Trying to delete {sequencing_run_path}")
 
-    def delete_flow_cell_directory(self) -> None:
+    def delete_sequencing_run_directory(self) -> None:
         """
-        Delete the flow cell directory if it fulfills all requirements.
+        Delete the sequencing runs and demultiplexed runs directory if it fulfills all requirements.
         Raises:
             CleanFlowCellFailedError when any exception is caught
```
Comment (ChrOertlin) [inline cg/services/illumina_services/illumina_clean_sequencing_run_service.py:62]: needs a rename

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 2]
```diff
@@ -14,54 +14,63 @@
...
-            if self.can_flow_cell_directory_be_deleted():
+            if self.can_sequencing_run_directory_be_deleted():
                 if self.dry_run:
-                    LOG.debug(f"Dry run: Would have removed: {self.flow_cell.path}")
+                    LOG.debug(f"Dry run: Would have removed: {self.seq_run_dir_data.path}")
                     return
-                remove_directory_and_contents(self.flow_cell.path)
+                remove_directory_and_contents(self.seq_run_dir_data.path)
         except Exception as error:
             raise CleanFlowCellFailedError(
```
Comment (ChrOertlin) [inline cg/services/illumina_services/illumina_clean_sequencing_run_service.py:72]: needs a rename

File: cg/cli/clean.py [hunk 3]
```diff
@@ -253,11 +255,11 @@ def hk_bundle_files(
...
 
 
-@clean.command("flow-cells")
+@clean.command("illumina-sequencing-runs")
 @DRY_RUN
 @click.pass_obj
-def clean_flow_cells(context: CGConfig, dry_run: bool):
-    """Remove flow cells from the flow cells and demultiplexed runs folder."""
+def clean_illumina_sequencing_runs(context: CGConfig, dry_run: bool):
+    """Remove raw and demultiplexed illumina sequencing runs from the from hasta."""
```
Comment (diitaz93) [inline cg/cli/clean.py]: I think we should stick to the agreed jargon of `sequencing` and `demultiplex`, `raw` may be unclear
```suggestion
    """Remove Illumina run sequencing and demultiplex data from the from hasta."""
```

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 4]
```diff
@@ -14,54 +14,63 @@
...
 from cg.utils.time import is_directory_older_than_days_old
 
 LOG = logging.getLogger(__name__)
 
 
-class CleanFlowCellAPI:
+class IlluminaCleanSequencingRunsService:
     """
-    Handles the cleaning of flow cells in the flow cells and demultiplexed runs directories.
+    Handles the cleaning of illumina sequencing runs in the sequencing-runs and demultiplexed-runs directories.
```
Comment (diitaz93) [inline cg/services/illumina_services/illumina_clean_sequencing_run_service.py:30]:
```suggestion
    Handles the cleaning of Illumina runs in the sequencing-runs and demultiplexed-runs directories.
```

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 5]
```diff
@@ -14,54 +14,63 @@
...
         self.status_db: Store = status_db
         self.hk_api: HousekeeperAPI = housekeeper_api
-        self.flow_cell = IlluminaRunDirectoryData(sequencing_run_path=flow_cell_path)
+        self.seq_run_dir_data = IlluminaRunDirectoryData(sequencing_run_path=sequencing_run_path)
         self.dry_run: bool = dry_run
-        LOG.info(f"Trying to delete {flow_cell_path}")
+        LOG.info(f"Trying to delete {sequencing_run_path}")
 
-    def delete_flow_cell_directory(self) -> None:
+    def delete_sequencing_run_directory(self) -> None:
```
Comment (diitaz93) [inline cg/services/illumina_services/illumina_clean_sequencing_run_service.py:56]: The current name gives the idea that it only deletes the sequencing dir
```suggestion
    def delete_run_directory(self) -> None:
```

File: cg/cli/clean.py [hunk 6]
```diff
@@ -253,11 +255,11 @@ def hk_bundle_files(
...
 
 
-@clean.command("flow-cells")
+@clean.command("illumina-runs")
 @DRY_RUN
 @click.pass_obj
-def clean_flow_cells(context: CGConfig, dry_run: bool):
-    """Remove flow cells from the flow cells and demultiplexed runs folder."""
+def clean_illumina_runs(context: CGConfig, dry_run: bool):
+    """Remove illumina sequencing and demultiplexed runs from hasta."""
```
Comment (diitaz93) [inline cg/cli/clean.py:262]:
```suggestion
    """Remove Illumina sequencing and demultiplexed runs from hasta."""
```

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    IlluminaSampleSequencingMetrics,
+)
+from cg.store.store import Store
+from cg.utils.files import remove_directory_and_contents
+from cg.utils.time import is_directory_older_than_days_old
+
+LOG = logging.getLogger(__name__)
+
+
+class IlluminaCleanSequencingRunsService:
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/illumina_services/illumina_clean_sequencing_run_service.py:28]: Going for `IlluminaCleanRunService` to stratify by device first

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    """
+    Handles the cleaning of illumina sequencing runs in the sequencing-runs and demultiplexed-runs directories.
+    Requirements for cleaning:
+            Sequencing run is older than 21 days
+            Sequencing run is backed up
+            Sequencing run is in StatusDB
+            Sequencing run has sequencing metrics in StatusDB
+            Sequencing run has fastq files in Housekeeper or
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/illumina_services/illumina_clean_sequencing_run_service.py:36]: These are accurate

File: cg/services/illumina_services/illumina_clean_sequencing_run_service.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    def has_fastq_files_for_samples_in_housekeeper(self) -> bool:
+        """Check if all samples on the sequencing run have fastq files in housekeeper."""
+        return bool(self.get_files_for_samples_on_flow_cell_with_tag(tag=SequencingFileTag.FASTQ))
+
+    def has_spring_files_for_samples_in_housekeeper(self) -> bool:
+        """Check if all samples on the sequencing run have SPRING files in housekeeper."""
+        return bool(self.get_files_for_samples_on_flow_cell_with_tag(tag=SequencingFileTag.SPRING))
+
+    def has_spring_meta_data_files_for_samples_in_housekeeper(self) -> bool:
+        """Check if all samples on the sequencing run have SPRING metadata files in housekeeper."""
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/illumina_services/illumina_clean_sequencing_run_service.py:130]: See below, if not all files have bundles in returns None and thus the bool will only be true if files are returned

```Python
 def get_files_for_samples_on_flow_cell_with_tag(self, tag: str) -> list[File] | None:
        """Return the files with the specified tag for all samples on a sequencing run."""
        sequencing_run: IlluminaSequencingRun = self.get_sequencing_run_from_status_db()
        bundle_names: list[str] = [
            metrics.sample.internal_id for metrics in sequencing_run.sample_metrics
        ]
        files: list[File] = []
        for bundle_name in bundle_names:
            try:
                files.extend(
                    self.hk_api.get_files_from_latest_version(
                        bundle_name=bundle_name, tags=[tag, self.sequencing_run_dir_data.id]
                    )
                )
            except HousekeeperBundleVersionMissingError:
                continue
        if not files:
            LOG.warning(
                f"No files with tag {tag} found on flow cell {self.sequencing_run_dir_data.id}"
            )
            return None
        return files
```

File: cg/cli/clean.py [hunk 10]
```diff
@@ -269,14 +271,14 @@ def clean_flow_cells(context: CGConfig, dry_run: bool):
         directories_to_check.extend(get_directories_in_path(path))
     for flow_cell_directory in directories_to_check:
         try:
-            clean_flow_cell_api = CleanFlowCellAPI(
-                flow_cell_path=flow_cell_directory,
+            clean_service = IlluminaCleanSequencingRunsService(
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/cli/clean.py:274]: I agree, we should do this, not in this PR though

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+
+    # THEN checking whether the sequencing run has been backed up is TRUE
+    assert is_backed_up
+
+
+def test_is_not_sequencing_run_backed_up(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test to check that a sequencing run has been backed up."""
+    # GIVEN a clean illumina clean sequencing runs service with a sequencing run that has not been backed up
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:72]:
```suggestion
    # GIVEN an Illumina clean sequencing runs service with a sequencing run that has not been backed up
```

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+
+    # THEN a sequencing run is retrieved
+    assert isinstance(sequencing_run, IlluminaSequencingRun)
+
+
+def test_is_sequencing_run_in_statusdb(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test to check that a sequencing run is in statusdb."""
+    # GIVEN a clean illumina clean sequencing runs service with a store that contains a sequencing run
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:46]:
```suggestion
    # GIVEN an Illumina clean sequencing runs service with a store that contains a sequencing run
```

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 13]
```diff
@@ -0,0 +1,439 @@
...
+
+    # THEN checking whether a sequencing run is in statusdb is TRUE
+    assert is_in_statusdb
+
+
+def test_is_sequencing_run_backed_up(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test to check that a sequencing run has been backed up."""
+    # GIVEN a clean illumina clean sequencing runs service with a sequencing run that has been backed up
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:59]:
```suggestion
    # GIVEN an Illumina clean sequencing runs service with a sequencing run that has been backed up
```

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 14]
```diff
@@ -0,0 +1,439 @@
...
+    # THEN checking whether a sequencing run has IlluminaSampleSequencingMetrics entries returns True
+    assert has_metrics
+
+
+def test_is_directory_older_than_21_days_pass(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test to check whether a directory is older than 21 days."""
+
+    # GIVEN a clean illumina clean sequencing runs service with a sequencing run that can be deleted
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:117]:
```suggestion
    # GIVEN an Illumina clean sequencing runs service with a sequencing run that can be deleted
```

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+    # THEN a IlluminaSampleSequencingMetrics entry is retrieved
+    assert isinstance(metrics[0], IlluminaSampleSequencingMetrics)
+
+
+def test_has_sequencing_metrics_in_statusdb(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test to check that a sequencing run has IlluminaSampleSequencingMetrics entries in statusdb."""
+
+    # GIVEN a clean illumina clean sequencing runs service with a store that contains a sequencing run with IlluminaSampleSequencingMetrics entries
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:103]:
```suggestion
    # GIVEN an Illumina clean sequencing runs service with a store that contains a sequencing run with IlluminaSampleSequencingMetrics entries
```

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 16]
```diff
@@ -0,0 +1,439 @@
...
+    # THEN checking whether a given sequencing run directory is older than 21 days is TRUE
+    assert is_older_that_21_days
+
+
+def test_is_directory_older_than_21_days_fail(
+    illumina_clean_service_can_be_removed: IlluminaCleanRunsService,
+):
+    """Test to check whether a directory is older than 21 days."""
+
+    # GIVEN a clean illumina clean sequencing runs service with a current time that is set to now.
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:138]:
```suggestion
    # GIVEN an Illumina clean sequencing runs service with a current time that is set to now.
```

File: tests/services/illumina_services/cleaning_services/test_clean_runs_service.py [hunk 17]
```diff
@@ -0,0 +1,439 @@
...
+        return_value=True,
+    ):
+        with mock.patch(
+            "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_fastq_files_for_samples_in_housekeeper",
+            return_value=False,
+        ):
+            with mock.patch(
+                "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_spring_meta_data_files_for_samples_in_housekeeper",
+                return_value=False,
+            ):
```
Comment (diitaz93) [inline tests/services/illumina_services/cleaning_services/test_clean_runs_service.py:313]: there is no need to recursively add contexts, you can add one patch after another separated by a comma forming only one patching environment
```suggestion
    with mock.patch(  "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.is_directory_older_than_21_days",
        return_value=True,
    ), with mock.patch(     "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_fastq_files_for_samples_in_housekeeper",
        return_value=False,
    ), with mock.patch(            "cg.services.illumina_services.cleaning_services.clean_runs_service.IlluminaCleanRunsService.has_spring_meta_data_files_for_samples_in_housekeeper",
        return_value=False,
    ):
```

**Review by diitaz93 — APPROVED** [review]
Looks good 👍

**Pull request merged**